File: src/main/java/ting/leetcode/java/arrays101/NumArray.java
package ting.leetcode.java.arrays101;

//303
public class NumArray {
    private final int[] prefixSumArr;

    public NumArray(int[] nums) {
        prefixSumArr = new int[nums.length + 1];
        int sum = 0;
        int index = 1;
        for (int num : nums) {
            sum += num;
            prefixSumArr[index++] = sum;
        }
    }

    public int sumRange(int left, int right) {
        return prefixSumArr[right + 1] - prefixSumArr[left];
    }
}

File: src/main/kotlin/ting/leetcode/arrays101/NumArray.kt
package ting.leetcode.arrays101

class NumArray(nums: IntArray) {
    private val prefixSumArr: IntArray = IntArray(nums.size + 1)

    init {
        var sum = 0
        for ((i, num) in nums.withIndex()) {
            sum += num
            prefixSumArr[i + 1] = sum
        }
    }

    fun sumRange(left: Int, right: Int): Int = prefixSumArr[right + 1] - prefixSumArr[left]
}
